components/sidebar/sidebar_category_list/sidebar_category_list.tsx

```diff
@@ -2,32 +2,260 @@
 // See LICENSE.txt for license information.
 
 import React from 'react';
+import {FormattedMessage} from 'react-intl';
+
+import {Spring, SpringSystem, util as MathUtil} from 'rebound';
+
+import {Channel} from 'mattermost-redux/types/channels';
+import {Team} from 'mattermost-redux/types/teams';
 
 import SidebarCategory from '../sidebar_category';
+import UnreadChannelIndicator from 'components/unread_channel_indicator';
 
 type Props = {
+    currentTeam: Team;
+    currentChannel: Channel | undefined;
     categories: any[];
+    unreadChannelIds: string[];
 };
 
 type State = {
-
+    showTopUnread: boolean;
+    showBottomUnread: boolean;
 };
 
+// scrollMargin is the margin at the edge of the channel list that we leave when scrolling to a channel.
+const scrollMargin = 15;
+
+// scrollMarginWithUnread is the margin that we leave at the edge of the channel list when scrolling to a channel so
+// that the channel is not under the unread indicator.
+const scrollMarginWithUnread = 60;
+
 export default class SidebarCategoryList extends React.PureComponent<Props, State> {
+    channelRefs: Map<string, HTMLDivElement>;
+    scrollbar: React.RefObject<HTMLDivElement>;
+    animate: SpringSystem;
+    scrollAnimation: Spring;
+
+    constructor(props: Props) {
+        super(props);
+
+        this.channelRefs = new Map();
+        this.state = {
+            showTopUnread: false,
+            showBottomUnread: false,
+        };
+        this.scrollbar = React.createRef();
+
+        this.animate = new SpringSystem();
+        this.scrollAnimation = this.animate.createSpring();
+        this.scrollAnimation.setOvershootClampingEnabled(true); // disables the spring action at the end of animation
+        this.scrollAnimation.addListener({onSpringUpdate: this.handleScrollAnimationUpdate});
+    }
+
+    componentDidUpdate(prevProps: Props) {
+        // reset the scrollbar upon switching teams
+        if (this.props.currentTeam !== prevProps.currentTeam) {
+            this.scrollbar.current!.scrollTop = 0;
+        }
+
+        // Scroll to selected channel so it's in view
+        if (this.props.currentChannel?.id !== prevProps.currentChannel?.id) { //eslint-disable-line no-undef
+            this.scrollToChannel(this.props.currentChannel!.id);
+        }
+
+        this.updateUnreadIndicators();
+    }
+
+    setChannelRef = (channelId: string, ref: HTMLDivElement) => {
+        if (ref) {
+            this.channelRefs.set(channelId, ref);
+        } else {
+            this.channelRefs.delete(channelId);
+        }
+    }
+
+    getFirstUnreadChannelFromChannelIdArray = (array: string[]) => {
+        if (this.props.currentChannel) {
```

Comment:
We could flip this if statement to reduce the nesting in this if statement to return early